### PR TITLE
Create new library for Hubspot API

### PR DESCRIPTION
**Pull request by blakewaud**
The `RPA.Hubspot` library currently focuses on getting data from the CRM module of Hubspot. The underlying python library `hubspot-api-client` is an automatically generated SDK from [OpenAPiTools](https://github.com/OpenAPITools/openapi-generator) and has many more commands available that I have not yet been able to wrap for RFW.

The biggest feature of this API is a "natural language" search mode for the search API whereby a RFW developer can write a query like so:

```robotframework
    ${contacts}=    Search for objects    CONTACTS
    ...    firstname    EQ    John    AND    lastname    EQ    Smith
```

Currently, the library returns found objects as the native object or list of those native objects. This allows a user to access many properties of the Hubspot objects via dot notation. Some of the underlying objects are dictionaries (most notably the `properties` attribute of `SimplePublicObject` types).

I will be updating a class-level docstring to describe all this before considering this PR no longer draft.

### CODE REVIEW

**Comment by blakewaud** [comment]
I have completed documentation and would like to get this looked at, @mikahanninen I know you had started to look at it, but not sure if you are able to complete the full review.

The tests use a `setup.py` file which defines the variables used by my test robot, not sure where to document that as that is currently ignored (my version includes my sandbox API key so I cannot commit it).

File: packages/main/tests/robot/test_hubspot.robot [hunk 1]
```diff
@@ -0,0 +1,170 @@
+*** Settings ***
+Documentation       API keys and variables must be provided in ./setup.py. These should use a live or
+...                 sandbox Hubspot environment to test the API, there is no mocking function.
+
+Library             RPA.Hubspot
+Library             Collections
+Variables           ./setup.py
```
Comment (mikahanninen) [inline packages/main/tests/robot/test_hubspot.robot:7]: I would call this `testvars.py` or something like that. The `setup.py` has special meaning in the Python packages and I would not like mix that into Robot Framework variables file.

The `testvars.py` can be added into `.gitignore` if that is only blocker related to renaming.

Comment (blakewaud) [reply packages/main/tests/robot/test_hubspot.robot:7]: Makes perfect sense, this is my first full fledged Python project that I've worked on and I was not aware of `setup.py` significance. I'll update that and switch `retry2` to `tenacity`.

File: packages/main/src/RPA/Hubspot.py [hunk 2]
```diff
@@ -0,0 +1,1654 @@
...
+
+        *** Keywords ***
+        Authorize Hubspot
+            ${secrets}=    Get secret    hubspot
+            Auth with api key    api_key    ${secrets}[API_KEY]
+
+    .. code-block:: python
+
+        from RPA.Hubspot import Hubspot
+        from RPA.Robocorp.Vault import RobocorpVault
```
[the remark this answers is not in the document]
Comment (mikahanninen) [reply packages/main/src/RPA/Hubspot.py:255]: There many similar imports in the keyword examples

**Review by mikahanninen — CHANGES_REQUESTED** [review]
In overall I really like how library is documented and tested, but keep in mind to check also that the generated API documentation looks what you expected.

Some fixes to be made, but otherwise good work 👍

File: packages/main/pyproject.toml [hunk 3]
```diff
@@ -82,9 +80,11 @@ java-access-bridge-wrapper = "^0.9.1"
 PyYAML = "^5.4.1"
 tenacity = "^8.0.1"
 htmldocx = "^0.0.6"
+hubspot-api-client = "^4.0.6"
+retry2 = "^0.9.3"
```
[the remark this answers is not in the document]
Comment (blakewaud) [reply packages/main/pyproject.toml]: Updated tests to use `testvars.py` and described documentation.

Replaced `retry2` with `tenacity`.

File: packages/main/src/RPA/Hubspot.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (blakewaud) [reply packages/main/src/RPA/Hubspot.py:260]: Fixed the import statement, I should have looked that up.

File: packages/main/src/RPA/Hubspot.py [hunk 4]
```diff
@@ -0,0 +1,1654 @@
...
+
+        vault = RobocorpVault()
+        secrets = vault.get_secret("hubspot")
```
[the remark this answers is not in the document]
Comment (blakewaud) [reply packages/main/src/RPA/Hubspot.py:263]: I did not know of this pattern, thanks!

**Review by mikahanninen — APPROVED** [review]

**Comment by blakewaud** [comment]
@mikahanninen Seems ready to go, all checks passed, but I would have expected the robot tests to fail since I haven't configured `testvars.py` in GitHub actions, does it not run the robot tests?

**Review by cmin764 — COMMENTED** [review]
Looks good and I feel you need to merge with latest master first so I can improve the diff/review experience.

A few other notes:
- It's best to make PRs from branches made in the same repo, not on forks. (ask for access if you can't push a branch into rpaframework)
- I guess the _buids/_ containing the wheel and tar.gz should be gitignored if not already or just removed. (if I haven't missed something meanwhile)
- I'll review it once again when the diffs shrinks and isolates the intention of this PR only, please notify me @blakewaud once this is ready again.

**Comment by cmin764** [comment]
> > Looks good and I feel you need to merge with latest master first so we can improve the diff/review experience.
> 
> @cmin764 I rebased against master 2 days ago (v13.0.2), I am not sure why the diffs are still so large...
> 
> > * It's best to make PRs from branches made in the same repo, not on forks. (ask for access if you can't push a branch into rpaframework)
> 
> When I started this, I did not have access, and I was not granted access until the project was half completed. Is there anything to do at this point?
> 
> > * I guess the _buids/_ containing the wheel and tar.gz should be gitignored if not already or just removed. (if I haven't missed something meanwhile)
> 
> I've removed these, they should not have been on this PR.

On the master rebasing, new stuff was added meanwhile:
```console
cmin@robocorp-mac rpaframework % git show
commit aa416b48ca6f2f4d3e916b9c15bcb13359595c2e (HEAD -> feature/hubspot-api, blakewaud/feature/hubspot-api)
cmin@robocorp-mac rpaframework % git merge master
Auto-merging poetry.lock
CONFLICT (content): Merge conflict in poetry.lock
Auto-merging packages/main/pyproject.toml
Auto-merging packages/main/poetry.lock
Automatic merge failed; fix conflicts and then commit the result.
```
and as you can see, there are even conflicts to be solved. You can solve them with `git mergetool -t <tool>` where the tool can be `vimdiff` and if you're not fluent in Vim, then you can go back into your IDE/Editor and search for the `<<<< HEAD` markers so you keep in your files both what was added recently + your intended changes and remove these markers for every conflict solved. (then once you quit the Vim 3-way-diff windows, you'll advance to the next file with conflicts to be solved -- we can setup a call for this if you want more guidance)

- Bonus: Use at any time `git merge --abort` to abort current merge and start over again using the above steps.
- Given the branching, it's all good with this PR, next time you can push a branch directly in the official repo ;). (this way we can work collaboratively over the same branches without adding each other's upstreams)

Good luck and let me know how it goes!

**Comment by blakewaud** [comment]
@cmin764 @mikahanninen Okay, so I had originally done a rebase on upstream/master@13.0.2 but when I then finished that rebase, I merged with this branch, thus confusing my history even more. I did not realize a merge from `upstream/master` would have been more appropriate in this instance. I have now completed an interactive rebase only selecting the correct commits from my original work and rebased them onto upstream/master@13.0.3. From what I see, the only way to truly fix the history on GitHub is via a force push. You are the only ones to have pulled down my work that I know of, so are you okay with me performing the force push?

**Comment by mikahanninen** [comment]
- On the `poetry.lock` conflict. It is totally ok to disregard `poetry.lock` file coming from **master**. You can just delete the file and recreate it in your branch by running `poetry lock` or `poetry update`. The lock file is generated file based on the definitions in the `pyproject.toml` file (so as long as `.toml` file is correct then there is nothing to worry about the `.lock` file).
- I would not do any force pushes. As you have limited number of new files I would just create new branch in this repository, do normal repository pull request and forget your forked branch. (I will check permissions to this repository).

**Pull request closed without merging**